Jb50188

etaishev opened the pull request:
Added the "rpm sign" build stage for signing rpm after build

The last commit with exports needed to redefine the RPM sign build step for customization.

martyone commented:
We are currently in the process of upgrading to Qt Creator 4.12 - conflicts are unavoidable :\ I will ping you when this is done and our master is ready for new features.

martyone commented:
> We are currently in the process of upgrading to Qt Creator 4.12 - conflicts are unavoidable :\ I will ping you when this is done and our master is ready for new features.

The master branch is now on 4.12, you can rebase.

etaishev commented:
> > We are currently in the process of upgrading to Qt Creator 4.12 - conflicts are unavoidable :\ I will ping you when this is done and our master is ready for new features.
> 
> The master branch is now on 4.12, you can rebase.

Updated PR

martyone commented:
Under https://github.com/sailfishos/sdk-setup/pull/179 we are also dropping implication of "make-install" step during "deploy --rsync". So there should be similar split done to "MerMb2RsyncDeployStep" here unless I missed something.

etaishev commented:
> I would prefer squashing the two commits
> 
> * Mer: Add show dialog options for MerMb2RpmBuildStep
> * Mer: Split the deploy step to build and install rpm
> 
> into single commit: "Mer: Split the deploy step to build and deploy RPM" and mention in the commit message that "The 'deploy' command of mb2 used to imply package creation when ran behind Qt Creator. Nowadays the 'package' command needs to be used explicitly."
> 
> There should be a checkbox in the "RPM Build" step's config widget to control the "Sign" option, and another checkbox under Options > Sailfish OS > General to control the default.
> 
> Why is the export needed more exactly?
> 
> Is it really needed to move MerDeployStepWidget into its own file? At the same time you are exporting another class (MerProcessStep) from the original file.

Move MerDeployStepWidget is small refactoring -> `merdeploysteps.h` is so big and mix UI and logic.

Adding export useful for us, so that we can implement own custom build step, but if you have other good ideas on how to insert a build step into the process I would be happy to see them

martyone commented:
> Move MerDeployStepWidget is small refactoring -> `merdeploysteps.h` is so big and mix UI and logic.

It's a tiny class, closely related to the others, so I would prefer keeping them together.

> Adding export useful for us, so that we can implement own custom build step, but if you have other good ideas on how to insert a build step into the process I would be happy to see them

Every exported class is a promise of API stability. This is something I am reluctant to do as the API is still awaiting a significant refactor. So please keep the exports on your side for the now.

May I review your custom build steps implemented on top of these exports? Maybe there are parts generic enough for inclusion on our side.

etaishev commented:
> May I review your custom build steps implemented on top of these exports? Maybe there are parts generic enough for inclusion on our side.

I add you to prive repo with my plugin.

martyone commented:
> I add you to prive repo with my plugin.

OK, I will check. Consider adding Ville as well so that we can discuss conveniently.

martyone commented on src/plugins/mer/merdeploysteps.cpp:
```diff
@@ -93,6 +94,8 @@ namespace {
 const int CONNECTION_TEST_CHECK_FOR_CANCEL_INTERVAL = 2000;
 
 const char MER_RPM_VALIDATION_STEP_SELECTED_SUITES[] = "MerRpmValidationStep.SelectedSuites";
+const char MER_RPM_BUILD_STEP_SIGN[] = "MerRpmBuildStep.SignEnable";
```
name and value would be better matching

```suggestion
const char MER_RPM_BUILD_STEP_SIGN_PACKAGES[] = "MerRpmBuildStep.SignPackages";
```

martyone commented on src/plugins/mer/merdeploysteps.cpp:
```diff
@@ -732,6 +737,23 @@ void MerMb2RpmBuildStep::doRun()
+    if (!MerProcessStep::fromMap(map))
+        return false;
+
+    m_enableSignPackages = map.value(MER_RPM_BUILD_STEP_SIGN, m_enableSignPackages).toBool();
```
ditto

```suggestion
    m_signPackages = map.value(MER_RPM_BUILD_STEP_SIGN_PACKAGES, m_signPackages).toBool();
```

martyone commented on src/plugins/mer/merdeploystepwidget.h:
```diff
@@ -0,0 +1,47 @@
+    class MerDeployStepWidget;
+} // namespace Ui
+
+class MerDeployStepWidget : public ProjectExplorer::BuildStepConfigWidget
```
I would prefer moving this back to merdeploysteps.h/cpp.

martyone commented on src/plugins/mer/merdeploystepwidget.h:
```diff
@@ -0,0 +1,47 @@
+    Ui::MerDeployStepWidget *m_ui;
+};
+
+class MerRpmBuildStepWidget : public MerDeployStepWidget
```
I would prefer implementing this in merdeploysteps.h/cpp

martyone commented on src/plugins/mer/merdeploystepwidget.h:
```diff
@@ -0,0 +1,47 @@
+signals:
+    void enableSign(bool sign);
+private:
+    QCheckBox *m_signCheckBox;
```
naming

```suggestion
    void setSignPackages(bool sign);
signals:
    void signPackagesChanged(bool sign);
private:
    QCheckBox *m_signPackagesCheckBox;
```

martyone commented on src/plugins/mer/mer.pro:
```diff
@@ -151,3 +151,5 @@ HEADERS += \
     mercmakebuildconfiguration.h \
     merqmakebuildconfiguration.h
+
+!isEmpty(MER_ENABLE_SIGN): DEFINES += MER_ENABLE_SIGN
```
What would be the reason to exclude the signing functionality?

martyone reviewed (commented):
Rebase is needed.

martyone commented on src/plugins/mer/merhardwaredevicewizardpages.cpp:
```diff
@@ -522,5 +525,117 @@ bool MerHardwareDeviceWizardSetupPage::validatePage()
+
+    m_keyPathChoother = new Utils::PathChooser;
+    m_keyPathChoother->setExpectedKind(PathChooser::Kind::File);
+    m_keyPathChoother->setPromptDialogTitle(tr("Select the public key file for the sign key"));
```
"Select a Trusted GnuPG Public Key"

martyone commented on src/plugins/mer/merhardwaredevicewizardpages.cpp:
```diff
@@ -522,5 +525,117 @@ bool MerHardwareDeviceWizardSetupPage::validatePage()
+    m_keyPathChoother = new Utils::PathChooser;
+    m_keyPathChoother->setExpectedKind(PathChooser::Kind::File);
+    m_keyPathChoother->setPromptDialogTitle(tr("Select the public key file for the sign key"));
+    m_keyPathChoother->setToolTip(tr("Public key for verify packages"));
```
"Select a trusted GnuPG public key"

martyone commented on src/plugins/mer/merhardwaredevicewizardpages.cpp:
```diff
@@ -522,5 +525,117 @@ bool MerHardwareDeviceWizardSetupPage::validatePage()
     return true;
 }
 
```
```suggestion

/*!
 * \class PackageKeyDeployer
 * \internal
 */

```

martyone commented on src/plugins/mer/merhardwaredevicewizardpages.cpp:
```diff
@@ -522,5 +525,117 @@ bool MerHardwareDeviceWizardSetupPage::validatePage()
+signals:
+    void finished(bool ok, const QString &errorString);
+};
+
```
```suggestion

/*!
 * \class MerHardwareDeviceWizardPackageKeyDeploymentPage
 * \internal
 */

```

martyone commented on share/qtcreator/sfdk/modules/20-building-mb2.json:
```diff
@@ -152,6 +152,21 @@
             "trDescription": "Do not update build-time dependencies unless required explicitly with 'build-requires pull'."
+        },
+        {
+            "name": "package-signing-user",
```
```suggestion
            "name": "package.signing-user",
```

Similarly in the other two cases.

martyone commented on src/plugins/mer/merbuildconfigurationaspect.cpp:
```diff
@@ -202,7 +202,10 @@ const QStringList MerBuildConfigurationAspect::s_allowedSfdkOptions{
     "no-fix-version",
-    "no-pull-build-requires"
+    "no-pull-build-requires",
+    "package-signing-user",
+    "package-signing-passphrase",
+    "package-signing-passphrase-file"
```
ditto

martyone reviewed (commented):
Typo in commit title "Sfdk: Add sign packeges options"

etaishev commented on src/plugins/mer/merhardwaredevicewizardpages.cpp:
```diff
@@ -522,5 +525,117 @@ bool MerHardwareDeviceWizardSetupPage::validatePage()
+                "|| mkdir -p %1 && chmod 0700 %1 && echo '"
+                    + QString::fromLocal8Bit(reader.data())
+                + "' >> %1/%2 && sdk-deploy-key %1/%2; rm %1/%2";
+        runner.run(command.arg(dirName, QFileInfo(keyFilePath).fileName()), sshParams);
```
With the "-" option, the sdk-deploy-key must wait for the key to be entered in stdin, right?

martyone replied:
yup, that's the common convention that you can pass '-' in place of a file name and the command will read stdin instead.

BTW, you *may* need to also somehow close the input channel after writeDataToProcess (you will see whether it hangs or not)

martyone commented on src/plugins/mer/merdeploysteps.cpp:
```diff
@@ -733,6 +745,23 @@ bool MerMb2RpmBuildStep::init()
+            packageUserSigningArgs.append(QStringLiteral("-c") + QtcProcess::quoteArgUnix("package-signing-passphrase-file=%1").arg(m_signPackagePassFile));
+
+        if (!packageUserSigningArgs.empty()) {
+            auto env = pp->environment();
+            env.appendOrSet(QLatin1String(Sfdk::Constants::MER_SSH_SFDK_OPTIONS), packageUserSigningArgs.join(' '));
+            pp->setEnvironment(env);
```
This looks really hacky so I checked a bit deeper and I realized all the signing related configuration should be implemented in MerBuildConfigurationAspect/Widget instead. What should remain here is just something like:

```
    auto aspect = buildConfiguuration()->aspect<MerBuildConfigurationAspect>();
    QTC_ASSERT(aspect, return false);
    if (aspect->isSignPackagesEnabled())
        rpmCommand.addArgs("--sign", CommandLine::Raw);
```

martyone commented on src/plugins/mer/merdeploysteps.h:
```diff
@@ -265,10 +277,31 @@ class MerDeployStepWidget : public ProjectExplorer::BuildStepConfigWidget
     Ui::MerDeployStepWidget m_ui;
 };
 
+class MerRpmBuildStepWidget : public MerDeployStepWidget
+{
+    Q_OBJECT
+public:
+    MerRpmBuildStepWidget(MerProcessStep *step, Sfdk::BuildEngine *engine);
+    void setSignPackages(bool sign);
+    void setSignPackageKey(const QString &keyFile);
+    void setSignPackagePass(const QString &passphrasefile);
+signals:
+    void signPackagesChanged(bool sign);
+    void signPackageKeyChanged(const QString &key);
+    void signPackagePassFileChanged(const QString &file);
+private:
+    void initKeyChooser();
+    void initKeyPassphraseChooser();
+    QCheckBox *m_signPackagesCheckBox;
+    QLineEdit *m_signPackageKeyEdit;
+    Utils::PathChooser *m_signPackageKeyPassFileChooser;
+    Sfdk::BuildEngine *m_engine;
+};
+
```
ditto, should be implemented in MerBuildConfigurationAspect/Widget instead.

etaishev replied:
For Rsync deploy step, signing is not necessary, but build configuration is common for all methods. Wouldn't that be weird?

martyone replied:
rsync deployment is rather a hack that needs to be used in cooperation with regular RPM deployment. Anyway it is the "package" step, not the "deploy" step what is controlled by those options, so it's all OK.

martyone commented on src/plugins/mer/merbuildconfigurationaspect.cpp:
```diff
@@ -120,6 +129,63 @@ class MerBuildConfigurationWidget : public ProjectExplorer::NamedWidget
+        Kit *kit = m_aspect->buildConfiguration()->buildSystem()->kit();
+        BuildEngine *engine = MerSdkKitAspect::buildEngine(kit);
+
+        const QString signKeysFolder = QDir::toNativeSeparators(engine->sharedConfigPath().toString())
+                                       + Sfdk::Constants::BUILD_ENGINE_HOST_GNUPG_PATH_POSTFIX;
+        m_ui->signPackageKeyLineEdit->setText(m_ui->signPackageKeyLineEdit->text().arg(signKeysFolder));
+
+        QStringList keys;
+        QDir keysDir(signKeysFolder);
+        for (const auto &keyInfo : keysDir.entryInfoList({PACKAGE_SIGN_KEY_FILE_FILTER}, QDir::Files))
+            keys << keyInfo.completeBaseName();
+
+        auto completer = new QCompleter(keys, this);
+        completer->setCaseSensitivity(Qt::CaseInsensitive);
+        m_ui->signPackageKeyLineEdit->setCompleter(completer);
+
+        m_ui->signPackageKeyPassFileChooser->setExpectedKind(PathChooser::Kind::File);
+        m_ui->signPackageKeyPassFileChooser->setPromptDialogTitle(tr("Select the passphrase file for the sign key"));
+        const QString toolTip = tr("Passphrase file for the sign key");
+        m_ui->signPackageKeyPassFileChooser->lineEdit()->setPlaceholderText(toolTip);
+        m_ui->signPackageKeyPassFileChooser->setToolTip(toolTip);
```
`signKeysFolder` is a location internal to the SDK. It is not a place the user should use as a primary storage for their keys - it will be removed during SDK uninstallation. The user should keep their keys stored on the regular GPG keyring and then:

> When the package.signing-user <name> option is set, sfdk should ensure that the corresponding private key is exported as ~/SailfishSDK/vmshare/gnupg/<name>.key prior to executing a command on the build engine.

(quoted from sailfishos/sdk-setup#179)

So what we should offer here to choose from is what we get with `gpg --list-secret-keys`. I would start with a combo box with items in formatted as `NAME <EMAIL> [SHORT_ID]`.

I think we have no other option than to pass the ID as `package.signing-user` (which should accept exactly the same values as gpg's `--local-user` option). NAME+EMAIL may not be unique. But it would be more user friendly when seen in build output, so it would be great if all of ID+NAME+EMAIL could be combined in argument to gpg's `--local-user` option somehow. The documentation does not mention the details, the implementation does not seem to allow that after brief check - maybe you spot a possibility there https://github.com/gpg/gnupg/blob/357ad9ae29677c1676b56d2b81282e2f78ec8040/common/userids.c#L41

etaishev replied:
So in sfdk we have to call `gpg --export-secret-keys <key-name>` on the developer host and then use it?

martyone replied:
yup

etaishev replied:
Hm, on Mac OS no `gpg`  by default and on Windows it's will be avaliabel only if running QtCreator under GitBash

martyone replied:
I wouldn't see it as granted that gpg is available even on Linux by default. This is not an issue. Want to gpg sign? - Install gpg first! :)

On Windows you mean it is not in PATH I believe. We can locate it the same way as we do for SSH - sitting in the same dir as git: https://github.com/sailfishos/sailfish-qtcreator/blob/master/src/plugins/projectexplorer/projectexplorer.cpp#L1992

etaishev replied:
Add search git location because gpg installed with git bash

martyone commented on src/plugins/mer/merbuildconfigurationwidget.ui:
```diff
@@ -54,6 +54,62 @@
     </widget>
    </item>
+   <item row="2" column="0" colspan="2">
+    <widget class="QGroupBox" name="signPackagesGroupBox">
```
It would be more natural if `Additional sfdk configuration:` came at the very end. I would also try to go without (visible) group box. Just something like:

```
RPM spec file:                 [                                            ] [Browse...]
[X] Sign packages
        Signing user:          [                                           v]
        Passphrase file:       [                                            ] [Browse...]
Additional sfdk configuration: [                                            ]

etaishev commented on src/tools/sfdk/script.cpp:
```diff
@@ -260,6 +262,32 @@ class JSBuildEngineExtension : public QObject
+        QString keyName;
+        QString keyPassFile;
+        // TODO: HACK ??
+        for (const auto &option : Configuration::effectiveState()) {
```
@martyone what do you think about exporting key here and getting the arguments?

martyone replied:
I think we should instead add

```
void BuildEngine::importGpgKey(const QString &id, const Utils::FilePath &passphraseFile, const QObject *context,
            const Functor<bool> &functor);
```

expose it on `JSBuildEngineExtension` with help of `execAsynchronous` and implement the rest in javascript:

```javascript
export function maybeImportSigningKey() {
    if (!configuration.isOptionSet("package.signing-user"))
        return [true, ""];

    var signingUser = configuration.optionArgument("package.signing-user");
    var signingPassphraseFile = configuration.isOptionSet("package.signing-passphrase-file")
        ? configuration.optionArgument("package.signing-passphrase-file")
        : "";
    if (!engine.importGpgKey(signingUser, signingPassphraseFile))
        return [false, qsTr("Failed to share GnuPG key with the build engine")];

    return [true, ""];
}
```

We should also think about being more verbose about the reasons for a failure - gpg may print something useful on stderr.

martyone commented on src/libs/sfdk/packagesignkeys.cpp:
```diff
@@ -0,0 +1,238 @@
+
+namespace Sfdk {
+
+class GpgRunner
```
This needs to be implemented a non blocking way. Check e.g. how docker or vboxmanage is used in {docker,vbox}virtualmachine.cpp.

etaishev replied:
Why is this needed? Why can't we do with a non-blocking completion wait in QEventLoop?
For few gpg calls looks like a excess mechanism.

martyone replied:
Sfdk API is meant to be 100% asynchronous - no blocking operation that could freeze the UI inadvertently. This is well supported internally, so the implementation overhead is too negligible to bother with the question whether this particular external command may or may not ever block our application. Use of nested event loops on library side is a asking for issues and should be avoided when possible.

martyone commented on src/libs/sfdk/signingutils.cpp:
```diff
@@ -0,0 +1,374 @@
+
+    arguments << "--export-secret-keys" << fingerprint;
+
+    auto runner = std::make_unique<GpgRunner>(arguments);
+    if (runner->process()->program().isEmpty()) {
+        m_lastError = GPG_NOT_FOUND_ERROR;
+        QTimer::singleShot(0, context, std::bind(functor, false));
+        return;
+    }
```
This is too late to check for this. We should never enter this function if gpg was not found. All GPG related UIs should be disabled if GPG executable is not available. Simply drop this and the other occurrences.

martyone commented on src/plugins/mer/merhardwaredevicewizardpages.cpp:
```diff
@@ -522,5 +529,153 @@ bool MerHardwareDeviceWizardSetupPage::validatePage()
+        Core::MessageManager::write(signUtils.error(), Core::MessageManager::Flash);
+
+    for (const auto &keyInfo : p->gpgKeyInfoList)
+        m_keyComboBox->addItem(keyInfo.name);
```
Here I would do similarly as I suggested for build settings: insert items in the form "NAME [FINGERPRINT]" to ensure usability in case of multiple keys for single NAME.

etaishev commented on src/libs/sfdk/signingutils.h:
```diff
@@ -0,0 +1,49 @@
+
+class BuildEngine;
+
+struct GpgKeyInfo
+{
+    QString name;
+    QString fingerprint;
+};
+
+typedef QVector<GpgKeyInfo> GpgKeyInfoList;
+
+class SigningUtilsPrivate;
+class SFDK_EXPORT SigningUtils : public QObject
+{
+    Q_OBJECT
+public:
+    explicit SigningUtils(QObject *parent = nullptr);
+    virtual ~SigningUtils();
+
+    void availableKeys(const QObject *context, const Functor<bool, const GpgKeyInfoList &> &functor);
+    void exportSecretKey(const QString &fingerprint,
+                         const Utils::FilePath &keyPassFile,
+                         const Utils::FilePath &outputDir,
+                         const QObject *context,
+                         const Functor<bool> &functor);
+    void exportPublicKey(const QString &fingerprint,
+                         const QObject *context,
+                         const Functor<bool, const QByteArray &> &functor);
+    QString error() const;
+
+protected:
+    SigningUtils(std::unique_ptr<SigningUtilsPrivate> &&d, QObject *parent);
+    std::unique_ptr<SigningUtilsPrivate> d_ptr;
+
+private:
+    Q_DECLARE_PRIVATE(SigningUtils);
+    QString m_error;
+};
```
If we take out Sfdk::Device::importGpgKey we lose the ability to set the key in MerHardwareDeviceWizardPackageKeyDeploymentPage when adding a device. I think this is more convenient than a separate step and we need to keep this feature

martyone replied:
I am not sure I follow what you mean, but as I see it most of the code that is now part of `PackageKeyDeployerPrivate::deploy` would be moved under `Sfdk::Device::importGpgKey(const QString &id, const QObject *context, const Functor<bool, QString> &functor)` which can use `SigningTools` internally.

`MerHardwareDeviceWizardPackageKeyDeploymentPage::deployKey()` would than directly do

```
const QString id = gpgKeyInfoList.at(m_keyComboBox->currentIndex()).fingerprint;
m_device->importGpgKey(id, this, [&](bool ok, const QString &errorString) {
    /* same as now */
})
```

martyone replied:
Now I see that `m_device` is a `MerHardwareDevice` and `Sfdk::HardwareDevice` does not exit yet at that point. - Is that what you meant?

This is not going to change easily :\ OK, let's create `static Sfdk::Device::importGpgKey(const QSsh::SshConnectionParameters &sshParameters, const QString &id, const QObject *context, const Functor<bool, QString> &functor)` as a workaround then and let `MerHardwareDeviceWizardPackageKeyDeploymentPage::deployKey()` do

```
const QString id = gpgKeyInfoList.at(m_keyComboBox->currentIndex()).fingerprint;
HardwareDevice::importGpgKey(m_device->sshParameters(), id, this,
        [&](bool ok, const QString &errorString) {
    /* same as now */
})
```

This will allow us to follow the long term objective of minimizing the amount of code in the Mer plugin and keeping as much as possible inside the Sfdk library. Please mark it with "TODO Make non-static".

martyone commented on src/plugins/mer/mergeneraloptionswidget.cpp:
```diff
@@ -86,6 +90,45 @@ MerGeneralOptionsWidget::MerGeneralOptionsWidget(QWidget *parent)
+    const bool isGpgAvailable = Sfdk::isGpgAvailable();
+    if (!isGpgAvailable) {
+        m_ui->signErrorLabel->setText(tr("GPG utility not found!"));
+        m_ui->signErrorLabel->setStyleSheet(QLatin1String("color: \"yellow\""));
```
hm, I don't see any "yellow" usage in src/plugins. Maybe "red"? More occurrences.

etaishev replied:
"red" looks like a fatal error. Without gpg, we can safely continue using SDK. Maybe should hide items if gpg is not available?

vigejolla replied:
Red indeed looks like a fatal error. Hiding is not good either, as the user should be given a hint that it might be a good idea to install gpg. Perhaps something like Theme::IconWarningColor.name(QColor::HexRgb) or something like that could be used?

martyone replied:
Without GPG, we can safely continue using SDK but we cannot use GPG and that's what this part of UI is about :)

Red highlight is what upstream QtC uses consistently. Check e.g. Options > Version Control - try to enter invalid path in any field.

Alternatively, we could also show the label using default color and show the warning icon in front of it.

And we should also think about adding path chooser here, which already provides the desired feedback, but I would prefer to postpone that as we should ideally merge this before the QtC upgrade work is started.

vigejolla replied:
> Alternatively, we could also show the label using default color and show the warning icon in front of it.

Indeed. Now I feel a bit silly as when I looked at what upstream QtC uses in such situations, I saw the warning icon. That's why I wrote about using IconWarningColor. I feel so stupid for not realizing that we could also use the icon!

martyone replied:
@etaishev it would be nice to fix this one yet

martyone commented on src/plugins/mer/merhardwaredevicewizardpages.cpp:
```diff
@@ -522,5 +533,194 @@ bool MerHardwareDeviceWizardSetupPage::validatePage()
+
+        bool ok;
+        QString error;
+        execAsynchronous(std::tie(ok, error),
+                         &HardwareDevice::importPublicGpgKey,
+                         device->sshParameters(),
+                         id);
```
Now I realized maybe we could create a temporary `Sfdk::HardwareDevice` instance just for the wizard purposes. I do not remember the details, so could you please try creating it like in `MerHardwareDeviceManager::onDeviceAddedOrUpdated` (just omit the `Sdk::addDevice` call) make `importPublicGpgKey` non-static and call it on the temporary instance? (If this works, it allows us moving more code from the wizard to the library easily, so it would be very nice! :))

martyone commented on src/plugins/mer/mergeneraloptionswidget.cpp:
```diff
@@ -86,6 +90,45 @@ MerGeneralOptionsWidget::MerGeneralOptionsWidget(QWidget *parent)
+    if (!ok)
+        Core::MessageManager::write(error, Core::MessageManager::Flash);
+
+    auto signingUsersModel = new SigningUserListModel(availableKeys);
+    m_ui->defaultSigningUserComboBox->setModel(signingUsersModel);
+    const int index = signingUsersModel->getIndexBy(MerSettings::signingUser());
+    m_ui->defaultSigningUserComboBox->setCurrentIndex(index);
```
With `GpgKeyInfo::toString/fromString` we could easily fill the combo box with strings and then get current with `GpgKeyIngo::fromString(combo->currentText())`. You know, the most reliable- and easiest to maintain- code is that one we never wrote, so if we can go without model... :)

martyone commented on src/tools/sfdk/script.cpp:
```diff
@@ -242,6 +242,19 @@ class JSBuildEngineExtension : public QObject
+        return withEngine([=](BuildEngine *engine) {
+            QString error;
+            bool ok;
+            execAsynchronous(std::tie(ok, error), std::mem_fn(&BuildEngine::importPrivateGpgKey), engine, signingUser, FilePath::fromString(signingPassphraseFile));
```
soooo long line :)

martyone commented on src/libs/sfdk/device.cpp:
```diff
@@ -115,6 +117,52 @@ Device::MachineType Device::machineType() const
+            return;
+        }
+
+        QString error;
+
+        SshRemoteProcessRunner runner;
+        QEventLoop loop;
+        connect(&runner, &SshRemoteProcessRunner::connectionError, Sdk::instance(), [&]() {
+            error = tr("Device connection error: %1").arg(runner.lastConnectionErrorString());
+            ok = false;
+            loop.quit();
+        });
+        connect(&runner, &SshRemoteProcessRunner::processClosed, Sdk::instance(), [&]() {
+            if (runner.processExitCode() != 0) {
+                error = tr("Deploy key error: %1").arg(QString::fromUtf8(runner.readAllStandardError()));
+                ok = false;
+            }
+            loop.quit();
+        });
+        connect(&runner, &SshRemoteProcessRunner::processStarted, [&runner, &keyData]() {
+            runner.writeDataToProcess(keyData);
+        });
+        runner.run("sdk-deploy-key -", sshParameters);
+        loop.exec();
+
+        if (context_)
+            functor(ok, error);
```
Edited suggested change - use `enqueueImmediate` instead of `enqueue`.

martyone commented on src/plugins/mer/mersigninguserselectiondialog.cpp:
```diff
@@ -0,0 +1,73 @@
+{
+    QList<GpgKeyInfo> availableKeys;
+    QString gpgErrorString;
+    if (Sfdk::isGpgAvailable(gpgErrorString)) {
```
Swap condition, return, omit `else`, declare and initialize `const availableKeys` later in one step.

martyone commented on src/plugins/mer/mersigninguserselectiondialog.cpp:
```diff
@@ -0,0 +1,73 @@
+        return GpgKeyInfo();
+    }
+
+    QStringList items = Utils::transform(availableKeys, [](const GpgKeyInfo &info) {
```
ditto, &GpgKeyInfo::toString, const items

martyone commented on src/plugins/mer/mersigninguserselectiondialog.cpp:
```diff
@@ -0,0 +1,73 @@
+        return GpgKeyInfo();
+
+    const int index = items.indexOf(selected);
+    if (index > -1 && index < availableKeys.count())
+        return availableKeys.at(index);
+
+    return GpgKeyInfo();
```
case for assert

```suggestion
    QTC_ASSERT(index > -1 && index < availableKeys.count(), return {});
    return availableKeys.at(index);
```

martyone commented on src/plugins/mer/mersigninguserselectiondialog.h:
```diff
@@ -0,0 +1,23 @@
+public:
+    explicit MerSigningUserSelectionDialog(QObject *parent = nullptr);
+    static Sfdk::GpgKeyInfo selectSigningUser();
+signals:
```
garbage

martyone commented on src/plugins/mer/mersigninguserselectiondialog.h:
```diff
@@ -0,0 +1,23 @@
+#define MERSIGNINGUSERSELECTIONDIALOG_H
+
+#include <sfdk/utils.h>
+#include <QAbstractListModel>
```
garbage

martyone reviewed (commented):
This looks quite right already! :)

Please try to check for and fix the following style issues everywhere:

- long lines (100 chars soft limit, 120 hard limit, exceptions apply :))
- `return {}` instead of `return FooBar()` wherever possible
- prefer implicit capture in lambdas
- argument list wrapping - prefer simple wrapping without aligning under opening bracket

Typo in commit message "packeges".

I never checked the distribution of changes into commits - is that right?

Great to see the temporary device instance could be used!

martyone commented on src/plugins/mer/merbuildconfigurationwidget.ui:
```diff
@@ -41,19 +41,126 @@
+   <item row="2" column="1">
+    <layout class="QHBoxLayout" name="signingUserLayout_2">
+     <item>
+      <layout class="QHBoxLayout" name="horizontalLayout">
```
How about this?

martyone reviewed (commented):
The whitespace code would still profit from some more love :) I will really appreciate if you improve it but I am also not going to delay the review anymore on that.

I would go with the following commit messages for the last three commits:

Sfdk: Allow to sign packages
Mer: Allow to sign packages
Mer: Allow to deploy public GPG keys to devices

Please add "Co-authored-by: Martin Kampas <martin.kampas@jolla.com>" footers to the last three commits.

Please ensure that the changes are properly tested. Given the schedule of upcoming work I am going to test it myself only later after it is integrated.

etaishev commented on src/libs/sfdk/device.cpp:
```diff
@@ -115,6 +119,40 @@ Device::MachineType Device::machineType() const
+        arguments.append("-");
+
+        auto runner = std::make_unique<RemoteProcessRunner>(sshParameters(),
+                "sdk-deploy-key",
```
I think it would be better to add an argument, such as `-k --key and -f --file` instead of `-`, to use via ssh
@martyone ?

martyone replied:
The acceptance of either file name or '-' as a positional argument follows the conventions used by coreutils, so I would keep it as it. We could make the argument optional though, reading from stdin whether it is '-' or no file is passed at all. And allow to pass multiple keys at once . Not necessarily, just an idea.

martyone reviewed (commented):
Looks beautiful now! Thank you!

martyone commented on src/plugins/mer/merbuildconfigurationaspect.cpp:
```diff
@@ -120,6 +134,84 @@ class MerBuildConfigurationWidget : public ProjectExplorer::NamedWidget
+        const bool isGpgAvailable = Sfdk::isGpgAvailable(&gpgErrorString);
+        if (!isGpgAvailable) {
+            m_ui->signErrorLabel->setText(gpgErrorString);
+            m_ui->signErrorIconLabel->setPixmap(Icons::CRITICAL.pixmap());
```
Wasn't the intention to let it look less critical? :) Maybe Icons::WARNING would be enough?

martyone reviewed (commented):
I think we should put the warning into its own row to avoid spacing issues. Localized strings may be much longer.

Great that you took emitDone into use after rebase :)

martyone commented on src/plugins/mer/mergeneraloptionswidget.ui:
```diff
@@ -242,42 +242,130 @@
+         </layout>
+        </item>
+        <item>
+         <layout class="QHBoxLayout" name="horizontalLayout_6">
```
This hbox seems needless now.

martyone commented on share/qtcreator/sfdk/modules/20-building-mb2.mjs:
```diff
@@ -95,3 +95,17 @@ export function mapCompilationDatabasePaths() {
+        return [false, qsTr("Failed to share GnuPG key with the build engine")];
+
+    return [true, ""];
+}
```
Hm, somehow this re-appeared.

martyone reviewed (approved):
LGTM. Please tell me when you feel it is properly tested (also WRT the changes to build steps) and I can start merging it. Thank you!

etaishev commented:
Also need to keep in mind the `sdk-deploy-key`

martyone commented:
> Also need to keep in mind the `sdk-deploy-key`

Hm, I somehow thought this was already fixed. OK, in order to not delay merging the rest of this PR, let's split the "Mer: Allow to deploy public GPG keys to devices" commit out to another PR on top of the current one. It is the last commit, so it should be as easy as in

git branch jb50188-1 jb50188
git branch -f jb50188 jb50188^

Then push and open another PR from jb50188-1

etaishev commented:
Also put the changes from `Sfdk::Device` into a separate commit in `jb50188-1`

martyone commented:
OK, great! Just the 'Co-authored-by' footer is now missing from the "Sfdk: Allow to sign packages" commit. Don't forget to open PR for the other branch.
